### DISCUSSION TITLE
Negative sampling in GraphSage

### DISCUSSION THREAD

**Discussion by RituPande** [danielegrattarola/spektral#334] [Q&A]
Hi Daniele,

I am not sure if I have correctly understood how unsupervised learning problem is to be formulated in GraphSage using negative sampling .
 
I have understood the concept of negative sampling in WordToVec  wherein the input to the model for training are as follows:
( context_word, positive_sample_word ) = 1
( context_word, negative_sample_word_1) = 0
    .
    .
   .
( context_word, negative_sample_word_k) = 0

The last layer of the model is a single logistic regression unit, and the model is trained using binary_crossentropy loss.

Following is  my understanding how negative sampling is to be implemented for GraphSage. Can you please check if i have understood it correctly:

1. GraphSage requires sampling for two tasks. First, to select the neighbors of each node that need to be aggregated t o come up with new feature representation
2. Second, pair of node v and correct neighbor u is labeled as 1 and k instances of node v , random node u~, labeled as 0. Therefore the training set has (k+1) * (total node size of the graph used for training) entries
3. Aggregation and update are  performed to create node embeddings for each node  v  using neighbors sampled in step-1
4. The last layer takes the dot product of v and u / u~ for positive and negative training samples respectively and feeds the score to a logistic regression unit.
5. The loss function used is binary-cross entropy 

So I will have two model. The first model is the GraphSage Model that takes as inputs the Graph nodes features and neighbors sampled in step 1. The number of training samples here will be equal to number of graph nodes that are trained, say n. The model aggregates  the nodes with sampled neighbors and updates them to create an embedding h. The second model will have  be the neighbor prediction model that will take as input n embeddings( h) for each node from the previous model and  n*(k+1)  training samples from step 2., performs dot product for each of the  n*(k+1)  training samples, feeds the output to logistic regression layer and uses binary cross entropy loss for back propagation

Is this understanding correct?

Thanks

Ritu
